core/src/main/scala/kafka/log/FileMessageSet.scala

```diff
@@ -126,26 +126,136 @@ class FileMessageSet private[kafka](@volatile var file: File,
+   */
+  def searchForTimestamp(targetTimestamp: Long, startingPosition: Int): Option[TimestampOffset] ={
+    var lastCheckedOffset = -1L
+    var maxTimestampChecked = Message.NoTimestamp
+    var foundOrAborted = false
+    def handleMessage(offset: Long, position: Int, magic: Byte, timestamp: Long): Boolean = {
+      lastCheckedOffset = offset
+      // Stop searching once we see message format before 0.10.0 or we found the message.
+      // This equivalent as treating message without timestamp has the largest timestamp.
+      // We do this to avoid scanning the entire log if no message has a timestamp.
+      if (magic == Message.MagicValue_V0) {
+        foundOrAborted = true
+        true
+      } else if (timestamp >= targetTimestamp) {
+        foundOrAborted = true
+        true
+        // Continue searching but record the last checked offset.
+      } else {
+        maxTimestampChecked = math.max(maxTimestampChecked, timestamp)
+        false
+      }
+    }
+
+    try scanAndHandleMessages(startingPosition, handleMessage)
+    catch {
+      case e: Exception => throw new KafkaException(s"Failed to search target timestamp $targetTimestamp due to", e)
+    }
+
+    if (foundOrAborted)
+      Some(TimestampOffset(maxTimestampChecked, lastCheckedOffset))
+    else if (lastCheckedOffset >= 0)
+      // we have finished scanning the log, the offset looked for is after the last checked offset.
+      Some(TimestampOffset(maxTimestampChecked, lastCheckedOffset + 1))
+    else
+      // We did not see any message
+      None
+  }
+
+  /**
+   * Return the largest timestamp of the messages after a given position in this file message set.
+   * @param startingPosition The starting position.
+   * @return The largest timestamp of the messages after the given position.
+   */
+  def largestTimestampAfter(startingPosition: Int): Long = {
+    var maxTimestamp = Message.NoTimestamp
+    def handleMessage(offset: Long, position: Int, magic: Byte, timestamp: Long): Boolean = {
+      maxTimestamp = math.max(maxTimestamp, timestamp)
+      false
+    }
+    try scanAndHandleMessages(startingPosition, handleMessage)
+    catch {
+      case e: Exception => throw new KafkaException(s"Failed to find the largest timestamp after position" +
+        s" $startingPosition due to ", e)
+    }
+    maxTimestamp
+  }
+
+  /**
+   * A helper function that does a linear scan over the log starting from a given position. A message handling function
+   * is passed to handle each message.
+   * @param startingPosition the starting physical position of the scan
+   * @param handleAndMaybeStop The message handling function. It takes (offset, position, magic, timestamp) as the arguments.
+   *                           This method should return true to stop the scan, return false to continue the scan.
+   */
+  private def scanAndHandleMessages(startingPosition: Int,
+                                    handleAndMaybeStop: (Long, Int, Byte, Long) => Boolean) {
```

Comment:
Good point. I think we can use the iterator for timestamp search. We probably still need to keep a separate scan function for offset search because the iterator does not give the position. But this will be only used by offset search so there is no function argument passed in.